fastai/text/data.py

```diff
@@ -13,44 +13,106 @@
 text_extensions = {'.txt'}
 
 class LanguageModelPreLoader(Callback):
-    "Transforms the texts in `dataset` in a stream for language modelling."
-    def __init__(self, dataset:LabelList, lengths:Collection[int]=None, bs:int=64, bptt:int=70, backwards:bool=False, 
-                 shuffle:bool=False, drop_last:bool=False):
-        self.dataset,self.bs,self.bptt,self.backwards = dataset,bs,bptt,backwards
-        self.shuffle,self.drop_last = shuffle,drop_last
-        self.lengths = np.array(ifnone(lengths, [len(o) for o in dataset.x.items]))
-        self.n = self.lengths.sum() // self.bs
-        
-    def __len__(self):
-        if self.item is not None:    return 1
-        if (self.n-1)%self.bptt == 0 or self.drop_last: return ((self.n-1) // self.bptt) * self.bs
-        return ((self.n-1) // self.bptt + 1) * self.bs
+    "Create a dataloader with bptt slightly changing."
     
+    class CircularIndex():
+        #When the index exceeds the length of self.idx then it is wrap to start at the head 
+        #or the end if indexing is moving backwards. The shuffled is done in-place, 
+        def __init__(self, length:int, forward:bool): 
+            self.idx      = np.arange(length)
+            self.forward_ = forward
+        def __getitem__(self, i): 
+            idx = self.idx
+            return idx[ i%len(idx) if self.forward_ else len(idx)-1-i%len(idx) ]
+        def __len__(self) -> int: return len(self.idx)
+        def shuffle(self): np.random.shuffle(self.idx)
+        def forward(self, forward:bool=True): self.forward_ = forward
+
+    def __init__(self, dataset:LabelList, bs:int=32, bptt:int=70, backwards:bool=False, shuffle:bool=False):
```

Comment:
I really liked the idea of being able to pass lengths, which would be computed only if not passed (since it can be long on a giant dataset). Could you add this to your PR?

Reply:
ok i will measure whether it makes the code faster

Reply:
This code show that we can save about 1.5 seconds for a very long ragged array (25 mio)

`lengths` is an 25 mio long array of lengths
`trainIDS_` is an 25 mio long array of arrays of tokens

def l1(lengths): 
...totalToks=0
...for l in lengths: totalToks += l
...return totalToks
def l2(items): 
...totalToks=0
...for rag in items: totalToks += len(rag)
...return totalToks    
print(f"number of rags:{len(trainIDS_)}")
%timeit l1(lengths)
%timeit l2(trainIDS_)

The result is:
> number of rags:24871686
> 2.45 s ± 55.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 3.94 s ± 27.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

It did make me go back and optimize some loops where i did not yet use local variable.
I have also modified the code to run with lengths=None and with a user provide lengths for the trainingdata.

However do not think the 1.5 seconds pr epoch justify the extra complexity of the argument list and the code.  What do you think ?